geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/set/SMoveDunitTest.java

```diff
@@ -0,0 +1,277 @@
+  }
+
+  @Test
+  public void givenBucketsMovedDuringSMove_thenOperationsAreNotLostOrDuplicated()
+      throws InterruptedException, ExecutionException {
+
+    final AtomicBoolean continueRunning = new AtomicBoolean(true);
+    final List<String> hashTags = getHashTagsForEachServer();
+
+    List<String> members1 = makeMemberList(10, "member1-");
+    List<String> members2 = makeMemberList(10, "member2-");
+    List<String> members3 = makeMemberList(10, "member3-");
+
+    jedis.sadd(hashTags.get(0) + KEY_1, members1.toArray(new String[] {}));
+    jedis.sadd(hashTags.get(1) + KEY_1, members2.toArray(new String[] {}));
+    jedis.sadd(hashTags.get(2) + KEY_1, members3.toArray(new String[] {}));
+
+    Future<Void> future1 = executor.runAsync(() -> repeatSMove(hashTags.get(0),
+        members1, continueRunning));
+    Future<Void> future2 = executor.runAsync(() -> repeatSMove(hashTags.get(1),
+        members2, continueRunning));
+    Future<Void> future3 =
+        executor.runAsync(() -> repeatSMoveWithSameSourceAndDest(hashTags.get(2),
+            members3, continueRunning));
+
+    for (int i = 0; i < 25 && continueRunning.get(); i++) {
+      clusterStartUp.moveBucketForKey(hashTags.get(i % hashTags.size()));
+      Thread.sleep(200);
+    }
+
+    continueRunning.set(false);
+
+    future1.get();
+    future2.get();
+    future3.get();
+  }
+
+  @Test
+  public void smove_isTransactional() {
+    IgnoredException.addIgnoredException(THROWING_CACHE_WRITER_EXCEPTION);
+
+    int primaryVMIndex = 1;
+    final String tag = "{" + clusterStartUp.getKeyOnServer("tag", primaryVMIndex) + "}";
+
+    final String sourceKey = tag + KEY_1;
+    int sourceSize = 2;
+    List<String> members = makeMemberList(sourceSize, "member1-");
+    jedis.sadd(sourceKey, members.toArray(new String[] {}));
+
+    String throwingKey = tag + "ThrowingRedisString";
+    int destinationSize = 2;
+    List<String> membersForThrowingKey = makeMemberList(2, "ThrowingRedisStringValue");
+    jedis.sadd(throwingKey, membersForThrowingKey.toArray(new String[] {}));
+
+    // Install a cache writer that will throw an exception if a key with a name equal to throwingKey
+    // is updated or created
+    clusterStartUp.getMember(1).invoke(() -> {
+      RedisClusterStartupRule.getCache()
+          .<RedisKey, RedisData>getRegion(DEFAULT_REDIS_REGION_NAME)
+          .getAttributesMutator()
+          .setCacheWriter(new SMoveDunitTest.ThrowingCacheWriter(throwingKey));
+    });
+
+    String memberToRemove = "member1-0";
+
+    assertThatThrownBy(
+        () -> jedis.smove(sourceKey, throwingKey, memberToRemove))
+            .hasMessage(SERVER_ERROR_MESSAGE);
+
+    // Assert smove has not happened
+    assertThat(jedis.scard(sourceKey)).isEqualTo(sourceSize);
+    assertThat(jedis.smembers(sourceKey)).contains(memberToRemove);
+    assertThat(jedis.scard(throwingKey)).isEqualTo(destinationSize);
+    assertThat(jedis.smembers(throwingKey)).doesNotContain(memberToRemove);
```

Comment:
It would be better to assert on the contents of the sets here rather than the sizes and whether or not they contain a specific element:
```
    assertThat(jedis.smembers(sourceKey)).containsExactlyInAnyOrderElementsOf(members);
    assertThat(jedis.smembers(throwingKey)).containsExactlyInAnyOrderElementsOf(membersForThrowingKey);
```